rls/rls-cargo-build-integration-notes.md

```diff
@@ -0,0 +1,105 @@
+For build system interop, Cargo and other build systems need to settle on a common
+abstraction (packages/crates/files dependencies? bare invocations?) in order to understand steps needed to compile a target built using the other system.
+
+(RLS) Who is responsible for build orchestration?
+1. RLS internally - needs to maintain its own version of abstract action graph and a mapping between it and a concrete build plan. Needs to understand which crate targets are dirty from a set of dirty files.
+2. Build system - RLS fetches the build plan from an external build system. Bonus points if the RLS can also only supply a list of dirty files and receive only the relevant sub-plan to execute.
+
+(Cargo <> other build system) What do systems need to know about each other?
+1. Build output for a given crate target - knowing this, a build system can only delegate to the other as a subroutine and use only the resulting artifacts (e.g. `.rlib`, debug symbols)
+2. Transitive file-level dependencies
+    1. Correct dependency/freshness tracking, e.g. external build system might need to understand that Cargo-managed crate target is fresh and can reuse (cloud) cache-managed artifacts
+    2. Generating external build system manifests, e.g. Makefiles
+3. Crate target dependencies - (together with file-level deps?) could be used to generate external build system manifests for more abstract build systems, e.g. define Buck/Bazel build rules
+
+# Current status
+## RLS
+RLS keeps its own crate target dependency graph with commands associated with them (which can be thought of as an action graph + build plan). To do so, it invokes Cargo to perform a full `check` build and intercepts every `rustc` invocaton and associates it with a crate target.  When modifying files the RLS detects dirty crate targets using a path heuristic and reruns the associated `rustc` commands.
+
+We'd like to:
+1. Not duplicate dep graph/build orchestration inside the RLS
+2. Understand other build systems - provide a common interface for an executable build plan
+
+### Build types
+#### Incremental
+We assume that the build plan has not changed, which means we can reuse it and query it for required commands that has to be rerun.
+  - triggerd by the RLS (source files changed)
+#### From scratch
+The build plan has possibly changed or is insufficient - we need to regenerate it using the build system.
+  - triggered by the user (e.g. build manifest or relevant package configuration was changed)
+
+## Cargo
+Cargo is capable of generating a build plan for `cargo build` (with `-Z unstable-options --build-plan`), which is a dependency graph between commands (`rustc` and calling build scripts, together with args, env, cwd), where each command is also tagged with a package name, version, crate target kind, Target/Host architecture differentiation and a list of resulting crate artifacts.
```

Comment:
Honestly there are bits that seem redundant and bits we'd like to know without having to parse the command args - this really depends on how much info we'd like to derive ourselves.

For example, simple {Target Host} enum doesn't tell us much - I'd imagine this might work with additional top-level explicit target and host triple, or if not, just include the actual target triple with invocation.

It's also useful to think what disambiguates the compile commands in the build plan.
If it's the hash of program, args, env, cwd and hash of deps, then we'd need to process these to guarantee consistency (e.g. parse and sort args).
However, if we'd like to disambiguate on all the associated metadata (which seems to be the only reason for package version, for example) then we'd need a lot more info, which means either relying on a build system to include all the metadata themselves (not realistic), or being able to parse the invocation and fill the missing metadata ourselves.

It might be worth rethinking what potential build plan consumers would like to know from it - the RLS would probably wind up parsing the command arguments to support simpler build systems; however other consumers might like to know more information up front.